protobuf_serialization/files/proto_parser.nim

```diff
@@ -1,232 +1,240 @@
-
-proc inlinecomment(): StringParser[string] =
-  ignorefirst(charmatch(Whitespace), s("/*") + allbut("*/") + s("*/")).repeat(1).map(combine)
-
-proc comment(): StringParser[string] = andor(endcomment(), inlinecomment()).repeat(1).map(combine)
-
-proc endstatement(): StringParser[string] =
-  ignoresides(inlinecomment(), ws(";"), comment())
-
-proc str(): StringParser[string] =
-  ignorefirst(inlinecomment(), optwhitespace(s("\"") + allbut("\"") + s("\""))).map(
-    proc(n: auto): string =
-      n[0][1]
-  ).ignorelast(comment())
-
-proc number(): StringParser[string] =
-  optwhitespace(charmatch(Digits)).onerror("Number doesn't match!")
-
-proc strip(input: string): string =
-  input.strip(true, true)
-
-proc token(): StringParser[string] =
-  ignoresides(comment(),
-    (
-      ignorefirst(charmatch(Whitespace), charmatch({'a'..'z', 'A'..'Z'})) +
-      ignorelast(optional(charmatch({'a'..'z', 'A'..'Z', '0'..'9', '_'})), charmatch(Whitespace))
-    ).map(combine)
-  , comment())
-
-proc token(name: string): StringParser[string] =
-  ignoresides(comment(), ws(name), comment()).map(strip)
-
-proc typespecifier(): StringParser[string] =
-  ignoresides(comment(),
-    (
-      optwhitespace(charmatch({'a'..'z', 'A'..'Z', '0'..'9', '_', '.'}))
-    )
-  , comment())
-
-proc syntaxline(): StringParser[string] = (token("syntax") + ws("=") + str() + endstatement()).map(
-  proc (stuple: auto): string =
-    stuple[0][1]
-)
-
-proc importstatement(): StringParser[ProtoNode] = (token("import") + str() + endstatement()).map(
-  proc (stuple: auto): ProtoNode =
-    ProtoNode(kind: Imported, filename: stuple[0][1])
-)
-
-proc package(): StringParser[string] = (token("package") + typespecifier() + endstatement()).map(
-  proc (stuple: auto): string =
-    stuple[0][1]
-)
-
-proc declaration(): StringParser[ProtoNode] = (optional(ws("repeated")) + typespecifier() + token() + ws("=") + number() + endstatement()).map(
-  proc (input: auto): ProtoNode =
-    result = ProtoNode(kind: Field, number: parseInt(input[0][1]), name: input[0][0][0][1], protoType: input[0][0][0][0][1], repeated: input[0][0][0][0][0] != "")
-)
-
-proc reserved(): StringParser[ProtoNode] =
-  (token("reserved") + (((number() + ws("to") + (number() / ws("max"))).ignorelast(ws(",")).map(
-    proc (input: auto): ProtoNode =
+type
+  TokenType = enum
+    Ident, Integer, Float, String, Symbol, ToImport
+
+  Token = ref object
+    typ: TokenType
+    text: string
+    index: int
+    line: int
+
+proc `$`(t: Token): string =
+  $t.typ & ": " & t.text
+
+proc `==`(t: Token, s: string): bool =
+  result = t.typ == Ident and t.text == s
+
+proc `==`(t: Token, tt: TokenType): bool =
+  t.typ == tt
+
+proc `==`(t: Token, c: char): bool =
+  t.typ == Symbol and t.text[0] == c
+
+proc tokenize(f: string): seq[Token] =
+  let lexer = peg(tokens, res: seq[Token]):
+    space    <- +Space
+    comment  <- "//" * *(1 - '\n')
+    comment2 <- "/*" * @"*/"
+    dblstring<- '"' * *(1 - '"') * '"'
```

Comment:
If I understand correctly, `"\""` will not be correctly tokenized.

Reply:
Yup, need to implement this https://developers.google.com/protocol-buffers/docs/reference/proto3-spec#string_literals